src/connection.cc

```diff
@@ -180,4 +181,129 @@ void IBConnection::flush(int64_t timeoutUsec) {
   // npkitCollectExitEvents(conn, NPKIT_EVENT_IB_SEND_EXIT);
 }
 
+// EthernetConnection
+
+EthernetConnection::EthernetConnection(Endpoint localEndpoint, Endpoint remoteEndpoint)
+    : stopRcvMessages_(false), abortFlag_(0) {
+  // Validating Transport Protocol
+  if (localEndpoint.transport() != Transport::Ethernet || remoteEndpoint.transport() != Transport::Ethernet) {
+    throw mscclpp::Error("Ethernet connection can only be made from Ethernet endpoints", ErrorCode::InvalidUsage);
+  }
+
+  // Instanciating Buffers
+  sendBuffer_ = new char[sendBufferSize_];
+  rcvBuffer_ = new char[rcvBufferSize_];
+
+  // Creating Thread to Accept the Connection
+  auto parameter = (getImpl(localEndpoint)->socket_).get();
+  std::thread t([this, parameter]() {
+    rcvSocket_ = std::make_unique<Socket>(nullptr, MSCCLPP_SOCKET_MAGIC, SocketTypeUnknown, abortFlag_);
+    rcvSocket_->accept(parameter);
+  });
+
+  // Starting Connection
+  sendSocket_ =
+      std::make_unique<Socket>(&(getImpl(remoteEndpoint)->socketAddress_), 0xdeadbeef, SocketTypeBootstrap, abortFlag_);
+  sendSocket_->connect();
+
+  // Ensure the Connection was Established
+  t.join();
+
+  // Starting Thread to Receive Messages
+  threadRcvMessages_ = std::thread(&EthernetConnection::rcvMessages, this);
+
+  INFO(MSCCLPP_NET, "Ethernet connection created");
+}
+
+EthernetConnection::~EthernetConnection() {
+  sendSocket_->close();
+  stopRcvMessages_ = true;
+  rcvSocket_->close();
+  threadRcvMessages_.join();
+}
+
+Transport EthernetConnection::transport() { return Transport::Ethernet; }
+
+Transport EthernetConnection::remoteTransport() { return Transport::Ethernet; }
+
+void EthernetConnection::write(RegisteredMemory dst, uint64_t dstOffset, RegisteredMemory src, uint64_t srcOffset,
+                               uint64_t size) {
+  // Validating Transport Protocol
+  validateTransport(dst, remoteTransport());
+  validateTransport(src, transport());
+
+  // Initializing Variables
+  char* srcPtr = reinterpret_cast<char*>(src.data()) + srcOffset / sizeof(char);
+  char* dstPtr = reinterpret_cast<char*>(dst.originalDataPtr()) + dstOffset / sizeof(char);
+  uint64_t sendSize = 0;
+
+  // Sending Info Data
+  sendSocket_->send(&dstPtr, sizeof(char*));
+  sendSocket_->send(&size, sizeof(uint64_t));
```

Comment:
Maybe we can create a structure as the header fields for ethernet data transfer. Then send these info at once.
Also we send data in sync manner, maybe async manner is more suitable for our case. (We don't need to implement this in this version)

Reply:
I had the same idea when I was implementing this method. I discussed this with Changho in that moment and he suggested to not create a struct because the time to parsing is long.